### ISSUE TITLE
Integration with Hyperledger

### ISSUE THREAD

**Issue by amiller** [initc3/HoneyBadgerMPC#24]
We need to commit to a plan for demonstrating the offline phase and online phase with an integration using hyperledger fabric.

**Comment by amiller** [comment]
A rough shot at a next steps TODO list based on conversation w/ sylvain:
- [x] Docker image that builds fabric and honeybadgermpc dependencies
- [ ] Identify a suitable "snip" point in Hyperledger Fabric endorser code, that lets endorsers communicate with the hbmpc application before approving a transaction. This should be at the endorser, roughly in between when it receives a transaction and when it signs it. Essentially the mpc will act like a "pre-ordering" service
- [x] Extend the fabric chaincode configuration file to include instructions for hbmpc to connect to endorsers. The main idea for a fabric integration will be to have the endorsers in a private chaincode deployment play the role of nodes in the MPC protocol. Fabric already provides a way to configure a set of endorsers which do not have communication requirements, so we will piggy-back on top of that.
 This will require forming point-to-point socket connections, to be taken care of from within the hbmpc application (the configuration file may just specify hostname, port, and authentication info)
- [ ] Clients submitting secret shared input. The idea will be for the client to send a private payload by initiating an instance of AVSS. We should try to use existing channels for client <--> endorser communication. Since the AVSS protocol is non-interactive for the dealer, the client should simply invoke an hbmpc script to generate a single message of the inputs. This is provided as private payload submitted along the transaction to each endorser.
   Given a transaction `T`, and payload shares `S[1],...,S[N]`, the client sends `(T,S[i])` to endorser `i`. The secret shares are imported in hbmpc using the AVSS protocol, and associated with the transaction hash. 
   When the transaction is (pre)-committed, the secret shared data is written to a key value store, `T.id => [S]`
- [ ] Initiating MPC computations. We'd like to use the chaincode to determine when to initiate a MPC computation. Initiating an MPC computation can include identifying which secret shared values to include as inputs (by their designated keys). MPC computations may take as input data from multiple parties, in which case the different inputs are referred to using the transaction IDs that provided them. The effect of the MPC is to write new values to the key-value store.
- [ ] Providing output. The output of an MPC computation can be either public, or transmitted to a client. If transmitted to a client, the client must request it from an endorser.

**Comment by amiller** [comment]
The following are suggestions from Angelo of where to start this:
https://hyperledger-fabric.readthedocs.io/en/release-1.2/systemchaincode.html
https://hyperledger-fabric.readthedocs.io/en/release-1.2/chaincode4noah.html
https://stackoverflow.com/questions/47134059/what-is-system-chaincode-in-fabric

A System Chaincode in fabric is what allows access to the underlying host, not just a container. This is where would want to patch in the ability to execute the Python hbmpc as a separate process with `subprocess`
 
[fabric/core/scc](https://github.com/hyperledger/fabric/tree/release-1.2/core/scc) - this folder contains all the available system chaincodes
and the infrastructure to manage them. Look also at [peer/node/start.go:611](https://github.com/hyperledger/fabric/blob/release-1.2/peer/node/start.go#L611)
to understand how the built-in SCC are instantiated.

**Comment by sbellem** [comment]
**Endorsement and validation** is no longer under [fabric/core/scc](https://github.com/hyperledger/fabric/tree/release-1.2/core/scc):

> The former system chaincodes for endorsement and validation have been replaced by the pluggable endorsement and validation function as described by the [Pluggable transaction endorsement and validation](https://hyperledger-fabric.readthedocs.io/en/latest/pluggable_endorsement_and_validation.html) documentation. 

The docs were outdated and just [got updated very recently](https://github.com/hyperledger/fabric/commit/316499ca5fd08ee6f2ee3302c400c64c4c07e689#diff-db95ae33651c43f7fd3b9dcb281f4844).

**Comment by sbellem** [comment]
@amiller @smkuls I have marked the docker image task as done, although we will probably make improvements and/or adjustments as the integration moves forward. I think it is good to get going right now though and that we can move to the other tasks. More details about the docker-based setup can be found in #52, more notably in https://github.com/initc3/HoneyBadgerMPC/issues/52#issuecomment-420336956 under the **Foreseeable possibe improvements/adjustments** paragraph.

**Comment by smkuls** [comment]
@hybridneo is working on this.

**Comment by amiller** [comment]
This is largely completed by @hybridNeo in separate repos https://github.com/hybridNeo/honeybadgerscc
Can you summarize in this issue what you did and talk about how we can merge it into this codebase so that we can continue on with it? Or how we can keep building on our separate fork of fabric?

**Comment by amiller** [comment]
@hybridNeo ping, please post somewhere a README or instructions on how we can run your RPS demo 
This repository is relevant but does not contain relevant instructions https://github.com/initc3/fabric

**Comment by amiller** [comment]
I think some other related repositories are:
- https://github.com/hybridNeo/honeybadgerscc  this is the system chaincode that defines the interface to hbmpc. script.sh builds it
- https://github.com/hybridNeo/fabric-test/tree/release-1.4 modification of a "deploy examples" script from fabric, that seems to be the main driver script used to launch the hbmpc exampes
- https://github.com/hybridNeo/rockpaperscissors example application implementing rock paper scissors
